provisioning/image_cache.go

```diff
@@ -60,27 +61,55 @@ func imageVolume() corev1.Volume {
 	}
 }
 
-func imageCacheConfig(targetNamespace string, config metal3iov1alpha1.ProvisioningSpec) (*metal3iov1alpha1.ProvisioningSpec, error) {
-	// The download URL looks something like:
-	// https://releases-art-rhcos.svc.ci.openshift.org/art/storage/releases/rhcos-4.2/42.80.20190725.1/rhcos-42.80.20190725.1-openstack.qcow2?sha256sum=123
-	downloadURL, err := url.Parse(config.ProvisioningOSDownloadURL)
+// Helper to transform the first level (metal3 pod) cache URLs to second level
+// (control-plane daemonset) cache
+func transformURL(targetNamespace, URL string) (string, error) {
+	downloadURL, err := url.Parse(URL)
 	if err != nil {
-		return nil, err
+		return "", err
 	}
 	imageName := path.Base(fileCompressionSuffix.ReplaceAllString(downloadURL.Path, ""))
 
-	// The first level of cache downloads the file and makes an uncompressed
-	// version of it available to this second-level cache somewhere like:
-	// http://metal3-state.openshift-machine-api:6180/images/rhcos-42.80.20190725.1-openstack.qcow2/rhcos-42.80.20190725.1-openstack.qcow2
+	// The first level of cache downloads the file and makes an version of it available
+	// to this second-level cache which then transforms it into the following format:
+	// Original format: http://<URL>:6180/images/rhcos-42.80.20190725.1-openstack.qcow2/cached-rhcos-42.80.20190725.1-openstack.qcow2
+	// Transformed format: http://metal3-state.openshift-machine-api:6181/images/rhcos-42.80.20190725.1-openstack.qcow2/cached-rhcos-42.80.20190725.1-openstack.qcow2
 	// See https://github.com/openshift/ironic-rhcos-downloader for more details
 	cacheURL := url.URL{
 		Scheme: "http",
 		Host: net.JoinHostPort(fmt.Sprintf("%s.%s.svc.cluster.local", stateService, targetNamespace),
 			baremetalHttpPort),
 		Path: fmt.Sprintf("/images/%s/%s", imageName, imageName),
 	}
-	config.ProvisioningOSDownloadURL = cacheURL.String()
-	return &config, nil
+	return cacheURL.String(), nil
+}
+
+func imageCacheConfig(targetNamespace string, config metal3iov1alpha1.ProvisioningSpec) (*string, error) {
+	// If the URL contains a comma-seperated list and are live images, split and transform the URLs
+	if getPreProvisioningOSDownloadURLs(&config) != nil {
```

Comment:
Maybe this function should just return a list of strings so we don't effectively have to write it twice.

Reply:
I think so. That make sense. I'll make the change.